File: src/backend/app/routes/acceptance_criteria.py
from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import Session

from app.db import get_session
from app.models_acceptance_criteria import (
    AcceptanceCriterionStatusUpdate,
    AcceptanceCriterionTextUpdate,
    ApplyApprovedRequest,
    RegenerateSelectedRequest,
    RegenerateSelectedResponse,
)
from app.models_conversation import AcceptanceCriterion, GenerateResult
from app.models_session import MessageRead
from app.repositories.acceptance_criteria import AcceptanceCriteriaRepository
from app.repositories.messages import MessageRepository
from app.repositories.sessions import SessionRepository
from app.repositories.uat_cases import UatCaseRepository
from app.services.acceptance_criteria import AcceptanceCriteriaService

router = APIRouter(
    prefix="/sessions/{session_id}/acceptance-criteria", tags=["acceptance-criteria"]
)


def get_acceptance_criteria_service(
    session: Session = Depends(get_session),
) -> AcceptanceCriteriaService:
    return AcceptanceCriteriaService(
        SessionRepository(session),
        AcceptanceCriteriaRepository(session),
        MessageRepository(session),
        UatCaseRepository(session),
    )


@router.get("", response_model=GenerateResult)
def list_acceptance_criteria(
    session_id: int, service: AcceptanceCriteriaService = Depends(get_acceptance_criteria_service)
) -> GenerateResult:
    items = service.list_items(session_id)
    if items is None:
        raise HTTPException(status_code=404, detail="Session not found")
    return GenerateResult(acceptance_criteria=items)


@router.patch("/{ac_id}", response_model=AcceptanceCriterion)
def update_acceptance_criterion_text(
    session_id: int,
    ac_id: int,
    data: AcceptanceCriterionTextUpdate,
    service: AcceptanceCriteriaService = Depends(get_acceptance_criteria_service),
) -> AcceptanceCriterion:
    updated = service.update_text(session_id, ac_id, data)
    if updated is None:
        raise HTTPException(status_code=404, detail="Acceptance criterion not found")
    return updated


@router.patch("/{ac_id}/status", response_model=AcceptanceCriterion)
def update_acceptance_criterion_status(
    session_id: int,
    ac_id: int,
    data: AcceptanceCriterionStatusUpdate,
    service: AcceptanceCriteriaService = Depends(get_acceptance_criteria_service),
) -> AcceptanceCriterion:
    updated = service.update_status(session_id, ac_id, data)
    if updated is None:
        raise HTTPException(status_code=404, detail="Acceptance criterion not found")
    return updated


@router.post("/{ac_id}/regenerate", response_model=RegenerateSelectedResponse)
def regenerate_selected(
    session_id: int,
    ac_id: int,
    data: RegenerateSelectedRequest,
    service: AcceptanceCriteriaService = Depends(get_acceptance_criteria_service),
) -> RegenerateSelectedResponse:
    result = service.regenerate_selected(session_id, ac_id, data)
    if result is None:
        raise HTTPException(status_code=404, detail="Acceptance criterion not found")
    return result


@router.post("/{ac_id}/apply-approved", response_model=GenerateResult)
def apply_approved(
    session_id: int,
    ac_id: int,
    data: ApplyApprovedRequest,
    service: AcceptanceCriteriaService = Depends(get_acceptance_criteria_service),
) -> GenerateResult:
    try:
        items = service.apply_approved(session_id, ac_id, data)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    if items is None:
        raise HTTPException(status_code=404, detail="Acceptance criterion not found")
    return GenerateResult(acceptance_criteria=items)


@router.post("/regenerate-all-kickoff", response_model=MessageRead, status_code=201)
def regenerate_all_kickoff(
    session_id: int, service: AcceptanceCriteriaService = Depends(get_acceptance_criteria_service)
) -> MessageRead:
    message = service.regenerate_all_kickoff(session_id)
    if message is None:
        raise HTTPException(status_code=404, detail="Session not found")
    return message

File: src/frontend/src/api/acceptanceCriteria.ts
import { request } from "./client";
import type { paths } from "./schema";

export type AcceptanceCriterion =
  paths["/sessions/{session_id}/acceptance-criteria"]["get"]["responses"]["200"]["content"]["application/json"]["acceptance_criteria"][number];
export type AcceptanceCriteriaList =
  paths["/sessions/{session_id}/acceptance-criteria"]["get"]["responses"]["200"]["content"]["application/json"];
type AcceptanceCriterionTextUpdate =
  paths["/sessions/{session_id}/acceptance-criteria/{ac_id}"]["patch"]["requestBody"]["content"]["application/json"];
type AcceptanceCriterionStatusUpdate =
  paths["/sessions/{session_id}/acceptance-criteria/{ac_id}/status"]["patch"]["requestBody"]["content"]["application/json"];
export type AcceptanceCriterionStatus = AcceptanceCriterionStatusUpdate["status"];
type RegenerateSelectedRequest =
  paths["/sessions/{session_id}/acceptance-criteria/{ac_id}/regenerate"]["post"]["requestBody"]["content"]["application/json"];
export type RegenerateSelectedResponse =
  paths["/sessions/{session_id}/acceptance-criteria/{ac_id}/regenerate"]["post"]["responses"]["200"]["content"]["application/json"];
type ApplyApprovedRequest =
  paths["/sessions/{session_id}/acceptance-criteria/{ac_id}/apply-approved"]["post"]["requestBody"]["content"]["application/json"];
export type RegenerateAllKickoffMessage =
  paths["/sessions/{session_id}/acceptance-criteria/regenerate-all-kickoff"]["post"]["responses"]["201"]["content"]["application/json"];

export const acceptanceCriteriaApi = {
  list: (sessionId: string): Promise<AcceptanceCriteriaList> =>
    request<AcceptanceCriteriaList>(`/sessions/${sessionId}/acceptance-criteria`),

  updateText: (
    sessionId: string,
    acId: number,
    data: AcceptanceCriterionTextUpdate,
  ): Promise<AcceptanceCriterion> =>
    request<AcceptanceCriterion>(`/sessions/${sessionId}/acceptance-criteria/${acId}`, {
      method: "PATCH",
      body: JSON.stringify(data),
    }),

  updateStatus: (
    sessionId: string,
    acId: number,
    status: AcceptanceCriterionStatus,
  ): Promise<AcceptanceCriterion> =>
    request<AcceptanceCriterion>(`/sessions/${sessionId}/acceptance-criteria/${acId}/status`, {
      method: "PATCH",
      body: JSON.stringify({ status } satisfies AcceptanceCriterionStatusUpdate),
    }),

  regenerateSelected: (
    sessionId: string,
    acId: number,
    data: RegenerateSelectedRequest,
  ): Promise<RegenerateSelectedResponse> =>
    request<RegenerateSelectedResponse>(
      `/sessions/${sessionId}/acceptance-criteria/${acId}/regenerate`,
      { method: "POST", body: JSON.stringify(data) },
    ),

  applyApproved: (
    sessionId: string,
    acId: number,
    candidates: AcceptanceCriterion[],
  ): Promise<AcceptanceCriteriaList> =>
    request<AcceptanceCriteriaList>(
      `/sessions/${sessionId}/acceptance-criteria/${acId}/apply-approved`,
      { method: "POST", body: JSON.stringify({ candidates } satisfies ApplyApprovedRequest) },
    ),

  regenerateAllKickoff: (sessionId: string): Promise<RegenerateAllKickoffMessage> =>
    request<RegenerateAllKickoffMessage>(
      `/sessions/${sessionId}/acceptance-criteria/regenerate-all-kickoff`,
      { method: "POST" },
    ),
};
